### ISSUE TITLE
Data race in Typer when compiling DottyPredef

### ISSUE THREAD

**Issue by DarkDimius** [lampepfl/dotty#528]
Can be reproduced by running `sbt testOnly dotc.tests` on this branch:
https://github.com/dotty-staging/dotty/tree/race-predef

```
[error] Test dotc.tests.dotty failed: java.lang.AssertionError: assertion failed: data race? overwriting symbol of dotty.DottyPredef$ / TypeRef(ThisType(TypeRef(NoPrefix,dotty)),DottyPredef$) / class dotty.tools.dotc.core.Types$CachedTypeRef / 9007 / 9748, took 1.276 sec
[error]     at scala.Predef$.assert(Predef.scala:165)
[error]     at dotty.tools.dotc.core.Types$NamedType.checkSymAssign(Types.scala:1263)
[error]     at dotty.tools.dotc.core.Types$NamedType.setDenot(Types.scala:1293)
[error]     at dotty.tools.dotc.core.Types$NamedType.withDenot(Types.scala:1286)
[error]     at dotty.tools.dotc.core.Types$TypeRef$.apply(Types.scala:1669)
[error]     at dotty.tools.dotc.core.Types$Type.select(Types.scala:819)
[error]     at dotty.tools.dotc.typer.Namer.moduleValSig(Namer.scala:602)
[error]     at dotty.tools.dotc.typer.Namer$Completer.typeSig(Namer.scala:465)
[error]     at dotty.tools.dotc.typer.Namer$Completer.completeInCreationContext(Namer.scala:499)
[error]     at dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:495)
[error]     at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:161)
[error]     at dotty.tools.dotc.core.SymDenotations$SymDenotation.info(SymDenotations.scala:143)
[error]     at dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:253)
[error]     at dotty.tools.dotc.core.SymDenotations$SymDenotation.flags(SymDenotations.scala:97)
[error]     at dotty.tools.dotc.core.SymDenotations$SymDenotation.is(SymDenotations.scala:119)
[error]     at dotty.tools.dotc.core.SymDenotations$SymDenotation.isStable(SymDenotations.scala:464)
[error]     at dotty.tools.dotc.core.Types$Type.widenIfUnstable(Types.scala:691)
[error]     at dotty.tools.dotc.core.Types$Type$$anonfun$member$1.apply(Types.scala:394)
[error]     at dotty.tools.dotc.core.Types$Type$$anonfun$member$1.apply(Types.scala:394)
[error]     at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
[error]     at dotty.tools.dotc.core.Types$Type.member(Types.scala:393)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typedIdent$1.wildImportRef$1(Typer.scala:192)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typedIdent$1.findRef$1(Typer.scala:240)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typedIdent$1.apply(Typer.scala:263)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typedIdent$1.apply(Typer.scala:79)
[error]     at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
[error]     at dotty.tools.dotc.typer.Typer.typedIdent(Typer.scala:79)
[error]     at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1017)
[error]     at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1076)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1088)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1086)
[error]     at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:147)
[error]     at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:51)
[error]     at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1086)
[error]     at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1125)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typedPackageDef$1.apply(Typer.scala:961)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typedPackageDef$1.apply(Typer.scala:960)
[error]     at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
[error]     at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:960)
[error]     at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1066)
[error]     at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1078)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1088)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1086)
[error]     at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:147)
[error]     at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:51)
[error]     at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1086)
[error]     at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1125)
[error]     at dotty.tools.dotc.typer.FrontEnd$$anonfun$typeCheck$1.apply$mcV$sp(FrontEnd.scala:41)
[error]     at dotty.tools.dotc.typer.FrontEnd.monitor(FrontEnd.scala:18)
[error]     at dotty.tools.dotc.typer.FrontEnd.typeCheck(FrontEnd.scala:39)
[error]     at dotty.tools.dotc.typer.FrontEnd$$anonfun$runOn$3.apply(FrontEnd.scala:52)
[error]     at dotty.tools.dotc.typer.FrontEnd$$anonfun$runOn$3.apply(FrontEnd.scala:52)
[error]     at scala.collection.immutable.List.foreach(List.scala:381)
[error]     at dotty.tools.dotc.typer.FrontEnd.runOn(FrontEnd.scala:52)
[error]     at dotty.tools.dotc.Run$$anonfun$compileSources$1$$anonfun$apply$mcV$sp$3.apply(Run.scala:55)
[error]     at dotty.tools.dotc.Run$$anonfun$compileSources$1$$anonfun$apply$mcV$sp$3.apply(Run.scala:52)
[error]     at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
[error]     at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
[error]     at dotty.tools.dotc.Run$$anonfun$compileSources$1.apply$mcV$sp(Run.scala:52)
[error]     at dotty.tools.dotc.Run$$anonfun$compileSources$1.apply(Run.scala:47)
[error]     at dotty.tools.dotc.Run$$anonfun$compileSources$1.apply(Run.scala:47)
[error]     at dotty.tools.dotc.util.Stats$.monitorHeartBeat(Stats.scala:68)
[error]     at dotty.tools.dotc.Run.compileSources(Run.scala:46)
[error]     at dotty.tools.dotc.Run.compile(Run.scala:33)
[error]     at dotty.tools.dotc.Driver.doCompile(Driver.scala:20)
[error]     at dotty.tools.dotc.Bench$.dotty$tools$dotc$Bench$$super$doCompile(Bench.scala:38)
[error]     at dotty.tools.dotc.Bench$$anonfun$doCompile$1.apply(Bench.scala:38)
[error]     at dotty.tools.dotc.Bench$$anonfun$doCompile$1.apply(Bench.scala:36)
[error]     at dotty.tools.dotc.Bench$$anonfun$ntimes$1.apply(Bench.scala:30)
[error]     at dotty.tools.dotc.Bench$$anonfun$ntimes$1.apply(Bench.scala:30)
[error]     at scala.collection.TraversableOnce$$anonfun$foldLeft$1.apply(TraversableOnce.scala:142)
[error]     at scala.collection.TraversableOnce$$anonfun$foldLeft$1.apply(TraversableOnce.scala:142)
[error]     at scala.collection.immutable.Range.foreach(Range.scala:166)
[error]     at scala.collection.TraversableOnce$class.foldLeft(TraversableOnce.scala:142)
[error]     at scala.collection.AbstractTraversable.foldLeft(Traversable.scala:104)
[error]     at scala.collection.TraversableOnce$class.$div$colon(TraversableOnce.scala:136)
[error]     at scala.collection.AbstractTraversable.$div$colon(Traversable.scala:104)
[error]     at dotty.tools.dotc.Bench$.ntimes(Bench.scala:30)
[error]     at dotty.tools.dotc.Bench$.doCompile(Bench.scala:36)
[error]     at dotty.tools.dotc.Driver.process(Driver.scala:31)
[error]     at dotty.tools.dotc.Bench$.dotty$tools$dotc$Bench$$super$process(Bench.scala:53)
[error]     at dotty.tools.dotc.Bench$$anonfun$process$1.apply(Bench.scala:53)
[error]     at dotty.tools.dotc.Bench$$anonfun$process$1.apply(Bench.scala:53)
[error]     at dotty.tools.dotc.Bench$$anonfun$ntimes$1.apply(Bench.scala:30)
[error]     at dotty.tools.dotc.Bench$$anonfun$ntimes$1.apply(Bench.scala:30)
[error]     at scala.collection.TraversableOnce$$anonfun$foldLeft$1.apply(TraversableOnce.scala:142)
[error]     at scala.collection.TraversableOnce$$anonfun$foldLeft$1.apply(TraversableOnce.scala:142)
[error]     at scala.collection.immutable.Range.foreach(Range.scala:166)
[error]     at scala.collection.TraversableOnce$class.foldLeft(TraversableOnce.scala:142)
[error]     at scala.collection.AbstractTraversable.foldLeft(Traversable.scala:104)
[error]     at scala.collection.TraversableOnce$class.$div$colon(TraversableOnce.scala:136)
[error]     at scala.collection.AbstractTraversable.$div$colon(Traversable.scala:104)
[error]     at dotty.tools.dotc.Bench$.ntimes(Bench.scala:30)
[error]     at dotty.tools.dotc.Bench$.process(Bench.scala:53)
[error]     at test.CompilerTest.compileArgs(CompilerTest.scala:159)
[error]     at test.CompilerTest.compileDir(CompilerTest.scala:104)
[error]     at dotc.tests.dotty(tests.scala:141)
```

**Comment by odersky** [comment]
It's a bootstrap problem. DottyPredef is already loaded when it gets defined again in the code.

**Comment by DarkDimius** [comment]
Fixed by #531 and #539
